File: backend/app/schemas/complaint.py
"""Pydantic schemas for Complaint CRUD operations."""
from __future__ import annotations
from datetime import date, datetime
from typing import Optional, List
from pydantic import BaseModel, Field


class ComplaintBase(BaseModel):
    source: Optional[str] = None
    customer_name: Optional[str] = None
    product_name: Optional[str] = None
    product_strength: Optional[str] = None
    batch_lot_number: Optional[str] = None
    affected_quantity: Optional[str] = None
    manufacturing_date: Optional[date] = None
    expiry_date: Optional[date] = None
    originating_site_block: Optional[str] = None
    impacted_npm: Optional[str] = None
    complaint_category: Optional[str] = None
    complaint_description: Optional[str] = None


class ComplaintCreate(ComplaintBase):
    """Schema for creating a new complaint (manual or AI-populated)."""
    status: Optional[str] = "draft"


class ComplaintUpdate(ComplaintBase):
    """Schema for partially updating a complaint."""
    status: Optional[str] = None


class AIAnalysisOut(BaseModel):
    id: str
    severity: Optional[str] = None
    suggested_next_action: Optional[str] = None
    initial_risk_assessment: Optional[str] = None
    regulatory_reportable: Optional[bool] = None
    root_cause_suggestions: Optional[List] = None
    capa_suggestions: Optional[List] = None
    confidence_scores: Optional[dict] = None
    model_used: Optional[str] = None
    created_at: datetime

    model_config = {"from_attributes": True}


class ComplaintOut(ComplaintBase):
    """Full complaint response schema."""
    id: str
    complaint_number: str
    status: str
    created_by: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    ai_analyses: Optional[List[AIAnalysisOut]] = []

    model_config = {"from_attributes": True}


class ComplaintListOut(BaseModel):
    """Paginated complaint list response."""
    items: List[ComplaintOut]
    total: int
    page: int
    page_size: int
    total_pages: int

File: frontend/src/types/complaint.ts
// TypeScript types mirroring backend Pydantic schemas

export interface User {
  id: string
  email: string
  name: string
  role: string
  created_at: string
}

export interface AIAnalysis {
  id: string
  severity?: 'Critical' | 'Major' | 'Minor'
  suggested_next_action?: string
  initial_risk_assessment?: string
  regulatory_reportable?: boolean
  root_cause_suggestions?: string[]
  capa_suggestions?: string[]
  confidence_scores?: Record<string, number>
  model_used?: string
  created_at: string
}

export interface Complaint {
  id: string
  complaint_number: string
  source?: string
  customer_name?: string
  product_name?: string
  product_strength?: string
  batch_lot_number?: string
  affected_quantity?: string
  manufacturing_date?: string
  expiry_date?: string
  originating_site_block?: string
  impacted_npm?: string
  complaint_category?: string
  complaint_description?: string
  status: ComplaintStatus
  created_by?: string
  created_at: string
  updated_at: string
  ai_analyses?: AIAnalysis[]
  risk_assessment?: Record<string, any>
  ai_analysis?: Record<string, any>
}

export type ComplaintStatus =
  | 'draft'
  | 'pending_triage'
  | 'ready_to_commit'
  | 'committed'
  | 'under_investigation'
  | 'capa_assigned'
  | 'closed'

export type SeverityLevel = 'Critical' | 'Major' | 'Minor'

export interface ComplaintDraft {
  source?: string
  customer_name?: string
  product_name?: string
  product_strength?: string
  batch_lot_number?: string
  affected_quantity?: string
  manufacturing_date?: string
  expiry_date?: string
  originating_site_block?: string
  impacted_npm?: string
  complaint_category?: string
  complaint_description?: string
}

export interface ComplaintListOut {
  items: Complaint[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

export interface AuditLogOut {
  id: string
  complaint_id: string
  actor: 'human' | 'ai'
  actor_name?: string
  field_name?: string
  old_value?: string
  new_value?: string
  action_type: string
  created_at: string
}
